modules/metrics-exporter-otlp/src/main/java/org/apache/ignite/internal/metrics/exporters/otlp/MetricProducer.java

```diff
@@ -0,0 +1,157 @@
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.export.CollectionRegistration;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableDoublePointData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableGaugeData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableHistogramData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableHistogramPointData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableLongPointData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableMetricData;
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Supplier;
+import org.apache.ignite.internal.lang.IgniteBiTuple;
+import org.apache.ignite.internal.metrics.DistributionMetric;
+import org.apache.ignite.internal.metrics.DoubleMetric;
+import org.apache.ignite.internal.metrics.IntMetric;
+import org.apache.ignite.internal.metrics.LongMetric;
+import org.apache.ignite.internal.metrics.Metric;
+import org.apache.ignite.internal.metrics.MetricProvider;
+import org.apache.ignite.internal.metrics.MetricSet;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Metrics producer which collect metrics in OpenTelemetry format.
+ */
+class MetricProducer implements CollectionRegistration {
+    private final MetricProvider metricsProvider;
+
+    private final Supplier<UUID> clusterIdSupplier;
+
+    private final String nodeName;
+
+    MetricProducer(MetricProvider metricsProvider, Supplier<UUID> clusterIdSupplier, String nodeName) {
+        this.metricsProvider = metricsProvider;
+        this.clusterIdSupplier = clusterIdSupplier;
+        this.nodeName = nodeName;
+    }
+
+    @Override
+    public Collection<MetricData> collectAllMetrics() {
+        IgniteBiTuple<Map<String, MetricSet>, Long> snapshot = metricsProvider.metrics();
+        @Nullable Long metricCount = snapshot.getValue();
+
+        if (metricCount == null || Objects.equals(metricCount, 0L)) {
+            return emptyList();
+        }
+
+        long epochNanos = Clock.getDefault().now();
+
+        Resource resource = Resource.builder()
+                .put("service.name", clusterIdSupplier.get().toString())
+                .put("service.instance.id", nodeName)
+                .build();
+
+        Collection<MetricData> result = new ArrayList<>(Math.toIntExact(metricCount));
+
+        for (MetricSet metricSet : snapshot.getKey().values())  {
+            InstrumentationScopeInfo scope = InstrumentationScopeInfo.builder(metricSet.name())
+                    .build();
+
+            for (Metric metric : metricSet) {
+                if (metric instanceof IntMetric) {
+                    IntMetric intMetric = (IntMetric) metric;
+
+                    result.add(ImmutableMetricData.createLongGauge(
```

Comment:
We've spent some effort to prepare the metrics infra, which is not producing significant amount of gargabe for GC. Looks like with this approach we will generate new objects on every metric push. At the same time, the list of metrics from call to call will be pretty stable. Doesn't this sdk has for example mutable APIs with prepared set of objects or anything like this?

Reply:
Added prepared set of wrappers under the real metrics.